model/author_category_topic.py

```diff
@@ -0,0 +1,161 @@
+                   word_topic_assignment: List[List[int]]):
+    """
+    Takes a set of documents and samples a new topic for each word within each document.
+    :param documents: a list of documents with their word ids
+    :param author_topic: a matrix describing the number of times each topic is used for each author
+    :param author_topic_c: a array of counts of authors
+    :param category_topic: a matrix describing the number of times each topic is used for each category
+    :param category_topic_c: a array of counts of categories
+    :param topic_word: a matrix describing the number of times each word within each topic
+    :param topic_word_c: the number of the times each topic is used within words
+    :param word_topic_assignment: A list of documents where each index is the given words topic
+
+
+
+    """
+    for d_index, doc in documents:
+        author = doc2author[d_index]
+        category = doc2category[d_index]
+        for w_index, word in enumerate(doc):
+            # Find the topic for the given word a decrease the topic count
+            topic = word_topic_assignment[d_index][w_index]
+            decrease_auth_cat(author, category, word, topic, author_topic, author_topic_c,
+                              category_topic, category_topic_c, topic_word, topic_word_c)
+
+            # Sample a new topic based on author_topic and topic word
+            # and assign it to the word we are working with
+            pz = _conditional_distribution_combination(author, category, word, author_topic, author_topic_c,
+                                                       category_topic, category_topic_c, topic_word, topic_word_c)
+            topic = np.random.multinomial(1, pz).argmax()
+            word_topic_assignment[d_index][w_index] = topic
+
+            # And increase the topic count
+            increase_auth_cat(author, category, word, topic, author_topic, author_topic_c,
+                              category_topic, category_topic_c, topic_word, topic_word_c)
+
+
+def triple_perplexity(documents: List[np.ndarray],
+                      author_topic, author_topic_c,
+                      category_topic, category_topic_c,
+                      topic_word, topic_word_c) -> float:
+    """
+    Calculates the perplexity based on the documents given
+    :param documents: a list of documents with word ids
+    :return: the perplexity of the documents given
+    """
+    n = 0
+    ll = 0.0
+    for d, doc in documents:
+        author = doc2author[d]
+        category = doc2category[d]
+        for w in doc:
+            ll += np.log(((topic_word[:, w] / topic_word_c) *
+                          (category_topic[category, :] / category_topic_c[category]) *
+                          (author_topic[author, :] / author_topic_c[author])).sum())
+            n += 1
+    return np.exp(ll / (-n))
+
+
+if __name__ == '__main__':
+    alpha = 0.1
+    beta = 0.1
+    iterationNum = 50
+    num_topics = 10
```

Comment:
```suggestion
    alpha = 0.01
    beta = 0.1
    iterationNum = 50
    num_topics = 90
```
Hvis jeg husker grid search værdierne korrekt.